[KIP-7] Applied KIP-13 (#16)
klaytn/kips: KIPs/kip-7-fungible_token.md

## Specification
<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Klaytn platforms (klaytn). -->
This document is heavily derived from [ERC-20](https://eips.ethereum.org/EIPS/eip-20) written by Fabian Vogelsteller and Vitalik Buterin.

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).

### Differences with ERC-20
This section describes the differences between KIP-7 and ERC-20. 

> - More optional functions are included (e.g. mint, burn and pause extensions)

Reviewer: ```suggestion
- More optional functions are included (e.g. mint, burn and pause extensions).
```

- Every token transfer/mint/burn MUST be tracked by event logs. This means that a Transfer event MUST be emitted for any action related to transfer/mint/burn.
- KIP-7 MUST implement [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) interface for each method group.

### KIP-13 Identifiers
The below table shows KIP-13 identifiers for interfaces defined in this proposal.

|Interface|KIP-13 Identifier|
|---|---|
|[IKIP7](#kip7-interface)|0x36372b07|
|[IKIP7Metadata](#metadata-extension)|0xa219a025|

> |[IKIP7Mint](#mint-extension)|0xeab83e20|

Reviewer 2: How about matching the KIP-17 with the interface name?
KIP-17 is naming extension interface names in this way "~able".

Author: Reflected. Thanks.

> |[IKIP7Mint](#mint-extension)|0xeab83e20|

Reviewer 2: ```suggestion
|[IKIP7Mint](#minting-extension)|0xeab83e20|
```

> |[IKIP7Burn](#burn-extension)|0x3b5a0bf8|

Reviewer 2: ```suggestion
|[IKIP7Burn](#burning-extension)|0x3b5a0bf8|
```

> |[IKIP7Pause](#pause-extension)|0x4d5507ff|

Reviewer 2: ```suggestion
|[IKIP7Pause](#pausing-extension)|0x4d5507ff|
```

}
```

### Metadata Extension
 
The **metadata extension** is OPTIONAL for KIP-7 smart contracts. 
This allows your smart contract to be interrogated for its name and for details about the assets which your token represent.

```solidity
/// @title KIP-7 Fungible Token Standard, optional metadata extension
///  Note: the KIP-13 identifier for this interface is 0xa219a025.
interface IKIP7Metadata {
    /// @dev Returns the name of the token.
    function name() external view returns (string memory);

    /// @dev Returns the symbol of the token, usually a shorter version of the
    /// name.
    function symbol() external view returns (string memory);

    /// @dev Returns the number of decimals used to get its user representation.
    /// For example, if `decimals` equals `2`, a balance of `505` tokens should
    /// be displayed to a user as `5,05` (`505 / 10 ** 2`).
    ///  Tokens usually opt for a value of 18, imitating the relationship between
    /// Ether and Wei.
    /// NOTE: This information is only used for _display_ purposes: it in
    /// no way affects any of the arithmetic of the contract, including

>     /// {IKIP-7-balanceOf} and {IKIP-7-transfer}.

Reviewer: ```suggestion
    /// `IKIP7.balanceOf` and `IKIP7.transfer`.
```

## Specification
<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Klaytn platforms (klaytn). -->
This document is heavily derived from [ERC-20](https://eips.ethereum.org/EIPS/eip-20) written by Fabian Vogelsteller and Vitalik Buterin.

### Comparing with ERC-20
Comparing with ERC-20 token standard, this token standard has some differences:

> - can include optional functions (e.g. mintable, burnable and pausable methods)

Reviewer: ```suggestion
-  More optional functions are included (e.g. mintable, burnable and pausable methods)
```

Author: reflected.

- Every token transfer/mint/burn MUST be tracked by event logs. This means that a Transfer event MUST be emitted for any action related to transfer/mint/burn.

> - MUST implement [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) interface for each method group.

Reviewer: ```suggestion
- The [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) interface for each method group MUST be implemented.
```

Author: reflected.

> ### KIP-7 Identifiers

Reviewer: ```suggestion
### KIP-13 Identifiers
```

The below table shows KIP-13 identifiers for interfaces defined in this proposal.

> |Interface|KIP-7 Identifier|

Reviewer: ```suggestion
|Interface|KIP-13 Identifier|
```

|---|---|

> |IKIP7|0xXXXXXXXX (TBD)|

Reviewer: When will this field be updated?

Author: I've updated already.

## Specification
<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Klaytn platforms (klaytn). -->
This document derived heavily from Ethereum's [ERC-20 token standard](https://eips.ethereum.org/EIPS/eip-20) written by Fabian Vogelsteller and Vitalik Buterin.
Comparing with ERC-20 token standard, this token standard has some differences:

> - can include optional functions (e.g. mint, burn, pausable methods)

Reviewer 2: ```suggestion
- can include optional functions (e.g. mintable, burnable and pausable methods)
```
How about matching form?
"mintable, burnable and pausable methods", "mint, burn and pause" or "mint, burn and pause methods"

|[transferFrom](#transferfrom)| |function transferFrom(address _from, address _to, uint256 _value) public returns (bool success)|
|[approve](#approve)| |function approve(address _spender, uint256 _value) public returns (bool success)|
|[allowance](#allowance)| |function allowance(address _owner, address _spender) public view returns (uint256 remaining)|

Every [KIP-7](https://klaytn.github.io/kips/KIPs/kip-7-fungible_token) compliant contract MUST implement [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) `supportsInterface` with the identifier `0xXXXXXXXX` for common methods.

> ```solidity

Reviewer: Could you please check out this? If you like it, please make this similar to it.

https://github.com/klaytn/kips/blob/dc5999a20b1746ff8b7dee49824c9b2adbfbfd5a/KIPs/kip-17-non_fungible_token.md#metadata-extension

Author: Ok. I will modify this like it.

Author: I updated. PTAL. :)

    function balanceOf(address account) external view returns (uint256);
    function transfer(address recipient, uint256 amount) external returns (bool);
    function allowance(address owner, address spender) external view returns (uint256);
    function approve(address spender, uint256 amount) external returns (bool);
    function transferFrom(address sender, address recipient, uint256 amount) external returns (bool);

    event Transfer(address indexed from, address indexed to, uint256 value);
    event Approval(address indexed owner, address indexed spender, uint256 value);
}
```

#### Detailed Methods (Optional)
|Name|Prototype|
|---|---|
|[name](#name)|function name() public view returns (string)|
|[symbol](#symbol)|function symbol() public view returns (string)|
|[decimals](#decimals)|function decimals() public view returns (uint8)|

Every [KIP-7](https://klaytn.github.io/kips/KIPs/kip-7-fungible_token) compliant contract with detailed methods MUST implement [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) `supportsInterface` with the identifier `0xXXXXXXXX` for detailed methods.
```solidity

> /// @title KIP-721 Non-Fungible Token Standard, optional detailed extension

Reviewer 2: ```suggestion
/// @title KIP-7 Fungible Token Standard, optional detailed extension
```

|[addMinter](#addminter)|function addMinter(address _account) public |
|[renounceMinter](#renounceminter)|function renounceMinter() public |

Every [KIP-7](https://klaytn.github.io/kips/KIPs/kip-7-fungible_token) compliant contract with mint methods MUST implement [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) `supportsInterface` with the identifier `0xXXXXXXXX` for mint methods.
```solidity

> /// @title KIP-721 Non-Fungible Token Standard, optional mint extension

Reviewer 2: ```suggestion
/// @title KIP-7 Fungible Token Standard, optional mint extension
```

|[burn](#burn)|function burn(uint256 _value) public |
|[burnFrom](#burnfrom)|function burnFrom(address _from, uint256 _value) public |

Every [KIP-7](https://klaytn.github.io/kips/KIPs/kip-7-fungible_token) compliant contract with burn methods MUST implement [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) `supportsInterface` with the identifier `0xXXXXXXXX` for burn methods.
```solidity

> /// @title KIP-721 Non-Fungible Token Standard, optional burn extension

Reviewer 2: ```suggestion
/// @title KIP-7 Fungible Token Standard, optional burn extension
```

|[addPauser](#addminter)|function addPauser(address _account) public |
|[renouncePauser](#renounceminter)|function renouncePauser() public |

Every [KIP-7](https://klaytn.github.io/kips/KIPs/kip-7-fungible_token) compliant contract with Pause methods MUST implement [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) `supportsInterface` with the identifier `0xXXXXXXXX` for Pause methods.
```solidity

> /// @title KIP-721 Non-Fungible Token Standard, optional pause extension

Reviewer 2: ```suggestion
/// @title KIP-7 Fungible Token Standard, optional pause extension
```

///  Note: the KIP-13 identifier for this interface is 0xXXXXXXXX.
interface IKIP7Pause {
    function supportsInterface(bytes4 interfaceID) external view returns (bool);    // KIP-13 interface

>     function transfer(address to, uint256 value) public whenNotPaused returns (bool);

Reviewer 2: The example of IKIP7Pause methods should be changed.

Author: I fixed. :) Thanks.

    function pause() public;

>     function unpause() public;

Reviewer 2: Thank you for changing :) Can you add isPauser, addPauser and renouncePauser?
